[BUG] "# => ALPHANUM" in word_delimiter filter's type_table can cause exception.

**Describe the bug**

If `# => ALPHANUM` is the only entry in a `word_delimiter` filter's `type_table`, analysis using that filter causes an exception. If the `#` is replaced with `\u0023` then it works as expected. The same behavior is seen with the `word_delimiter_graph` filter.

**To Reproduce**
```
curl "http://localhost:9200/_analyze?pretty" \
    -s \
    -X POST \
    -H 'Content-Type: application/json' \
    -d'
{
    "tokenizer": "whitespace",
    "filter": [
        {
            "type": "word_delimiter",
            "split_on_numerics": false,
            "type_table": [
                "# => ALPHANUM"
            ]
        }
    ],
    "text": "text1 #text2"
}
'
```

OS 2.6.0 (and 2.5.0 where I originally noticed) produces:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "no_such_element_exception",
        "reason" : null
      }
    ],
    "type" : "no_such_element_exception",
    "reason" : null
  },
  "status" : 500
}
```

Server log:
```
opensearch-node  | [2023-03-29T15:20:30,600][WARN ][r.suppressed             ] [opensearch-docker-node] path: /_analyze, params: {pretty=}
opensearch-node  | org.opensearch.transport.RemoteTransportException: [opensearch-docker-node][172.31.0.2:9300][indices:admin/analyze[s]]
opensearch-node  | Caused by: java.util.NoSuchElementException
opensearch-node  | 	at java.util.TreeMap.key(TreeMap.java:1602) ~[?:?]
opensearch-node  | 	at java.util.TreeMap.lastKey(TreeMap.java:298) ~[?:?]
opensearch-node  | 	at org.opensearch.analysis.common.WordDelimiterTokenFilterFactory.parseTypes(WordDelimiterTokenFilterFactory.java:156) ~[?:?]
opensearch-node  | 	at org.opensearch.analysis.common.WordDelimiterTokenFilterFactory.<init>(WordDelimiterTokenFilterFactory.java:89) ~[?:?]
opensearch-node  | 	at org.opensearch.index.analysis.AnalysisRegistry.getComponentFactory(AnalysisRegistry.java:156) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.index.analysis.AnalysisRegistry.buildCustomAnalyzer(AnalysisRegistry.java:285) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.admin.indices.analyze.TransportAnalyzeAction.buildCustomAnalyzer(TransportAnalyzeAction.java:238) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.admin.indices.analyze.TransportAnalyzeAction.analyze(TransportAnalyzeAction.java:165) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.admin.indices.analyze.TransportAnalyzeAction.shardOperation(TransportAnalyzeAction.java:151) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.admin.indices.analyze.TransportAnalyzeAction.shardOperation(TransportAnalyzeAction.java:88) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.support.single.shard.TransportSingleShardAction.lambda$asyncShardOperation$0(TransportSingleShardAction.java:131) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:73) [opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.action.ActionRunnable$2.doRun(ActionRunnable.java:88) ~[opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:806) [opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) [opensearch-2.6.0.jar:2.6.0]
opensearch-node  | 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
opensearch-node  | 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
opensearch-node  | 	at java.lang.Thread.run(Thread.java:833) [?:?]
```

Notes:

1. This only happens if `# => ALPHANUM` is the only entry in the `type_table`. Adding a second entry causes the analysis to succeed, but it behaves as if the `# => ALPHANUM` is not present. For example:
```
curl "http://localhost:9200/_analyze?pretty" \
    -s \
    -X POST \
    -H 'Content-Type: application/json' \
    -d'
{
    "tokenizer": "whitespace",
    "filter": [
        {
            "type": "word_delimiter",
            "split_on_numerics": false,
            "type_table": [
                "# => ALPHANUM".
                "@ => ALPHANUM"
            ]
        }
    ],
    "text": "text1 #text2"
}
'
```

Produces (incorrect):
```
{
  "tokens" : [
    {
      "token" : "text1",
      "start_offset" : 0,
      "end_offset" : 5,
      "type" : "word",
      "position" : 0
    },
    {
      "token" : "text2",
      "start_offset" : 7,
      "end_offset" : 12,
      "type" : "word",
      "position" : 1
    }
  ]
}
```

Note the `text2` token should be `#text2`. The ordering of the the two `type_table` entries does not matter.

2. Changing `# => ALPHANUM` to `\\u0023 => ALPHANUM` (0x23 is of course the character code for `#`) causes the analysis to work as expected. This:
```
curl "http://localhost:9200/_analyze?pretty" \
    -s \
    -X POST \
    -H 'Content-Type: application/json' \
    -d'
{
    "tokenizer": "whitespace",
    "filter": [
        {
            "type": "word_delimiter",
            "split_on_numerics": false,
            "type_table": [
                "\\u0023 => ALPHANUM"
            ]
        }
    ],
    "text": "text1 #text2"
}
'
```

Produces (correct):
```
{
  "tokens" : [
    {
      "token" : "text1",
      "start_offset" : 0,
      "end_offset" : 5,
      "type" : "word",
      "position" : 0
    },
    {
      "token" : "#text2",
      "start_offset" : 6,
      "end_offset" : 12,
      "type" : "word",
      "position" : 1
    }
  ]
}
```

**Expected behavior**

`# => ALPHANUM` in a `word_delimiter` or `word_delimiter_graph` filter's `type_table` should treat `#` as an alphanumeric character.

**Plugins**

None.

**Screenshots**

N/A

**Host/Environment (please complete the following information):**

Official OS 2.6.0 Docker image run under Docker Desktop 4.17.0 (99724) on Mac OS 13.2.1 (Intel).

**Additional context**

None

--- a/modules/analysis-common/src/main/java/org/opensearch/analysis/common/MappingCharFilterFactory.java
+++ b/modules/analysis-common/src/main/java/org/opensearch/analysis/common/MappingCharFilterFactory.java
@@ -54,7 +54,7 @@ public class MappingCharFilterFactory extends AbstractCharFilterFactory implemen
     MappingCharFilterFactory(IndexSettings indexSettings, Environment env, String name, Settings settings) {
         super(indexSettings, name);
 
-        List<MappingRule<String, String>> rules = Analysis.parseWordList(env, settings, "mappings", this::parse, false);
+        List<MappingRule<String, String>> rules = Analysis.parseWordList(env, settings, "mappings", this::parse);
         if (rules == null) {
             throw new IllegalArgumentException("mapping requires either `mappings` or `mappings_path` to be configured");
         }

--- a/modules/analysis-common/src/test/java/org/opensearch/analysis/common/BaseWordDelimiterTokenFilterFactoryTestCase.java
+++ b/modules/analysis-common/src/test/java/org/opensearch/analysis/common/BaseWordDelimiterTokenFilterFactoryTestCase.java
@@ -211,8 +211,8 @@ public abstract class BaseWordDelimiterTokenFilterFactoryTestCase extends OpenSe
     }
 
     public void testTypeTableParsingError() {
-        String[] rules = { "# This is a comment", "$ => DIGIT", "\\u200D => ALPHANUM", "abc => ALPHA" };
+        String[] rules = { "# This is a comment", "# => ALPHANUM", "$ => DIGIT", "\\u200D => ALPHANUM", "abc => ALPHA" };
         RuntimeException ex = expectThrows(RuntimeException.class, () -> createTokenFilterFactoryWithTypeTable(rules));
-        assertEquals("Line [4]: Invalid mapping rule: [abc => ALPHA]. Only a single character is allowed.", ex.getMessage());
+        assertEquals("Line [5]: Invalid mapping rule: [abc => ALPHA]. Only a single character is allowed.", ex.getMessage());
     }
 }

--- a/modules/analysis-common/src/test/java/org/opensearch/analysis/common/MappingCharFilterFactoryTests.java
+++ b/modules/analysis-common/src/test/java/org/opensearch/analysis/common/MappingCharFilterFactoryTests.java
@@ -37,6 +37,7 @@ public class MappingCharFilterFactoryTests extends OpenSearchTestCase {
 
     public void testRulesOk() throws IOException {
         MappingCharFilterFactory mappingCharFilterFactory = (MappingCharFilterFactory) create(
+            "# This is a comment",
             "# => _hashtag_",
             ":) => _happy_",
             ":( => _sad_"
@@ -64,7 +65,10 @@ public class MappingCharFilterFactoryTests extends OpenSearchTestCase {
     }
 
     public void testRulePartError() {
-        RuntimeException ex = expectThrows(RuntimeException.class, () -> create("# => _hashtag_", ":) => _happy_", "a:b"));
-        assertEquals("Line [3]: Invalid mapping rule : [a:b]", ex.getMessage());
+        RuntimeException ex = expectThrows(
+            RuntimeException.class,
+            () -> create("# This is a comment", "# => _hashtag_", ":) => _happy_", "a:b")
+        );
+        assertEquals("Line [4]: Invalid mapping rule : [a:b]", ex.getMessage());
     }
 }

--- a/modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/40_token_filters.yml
+++ b/modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/40_token_filters.yml
@@ -156,6 +156,7 @@
               - type: word_delimiter
                 split_on_numerics: false
                 type_table:
+                  - "# This is a comment"
                   - "# => ALPHANUM"
     - length: { tokens: 2 }
     - match: { tokens.0.token: text1 }
@@ -176,6 +177,7 @@
               - type: word_delimiter
                 split_on_numerics: false
                 type_table:
+                  - "# This is a comment"
                   - "# => ALPHANUM"
                   - "@ => ALPHANUM"
     - length: { tokens: 2 }
@@ -321,6 +323,7 @@
               - type: word_delimiter_graph
                 split_on_numerics: false
                 type_table:
+                  - "# This is a comment"
                   - "# => ALPHANUM"
     - length: { tokens: 2 }
     - match: { tokens.0.token: text1 }
@@ -341,6 +344,7 @@
               - type: word_delimiter_graph
                 split_on_numerics: false
                 type_table:
+                  - "# This is a comment"
                   - "# => ALPHANUM"
                   - "@ => ALPHANUM"
     - length: { tokens: 2 }

--- a/modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/40_token_filters.yml
+++ b/modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/40_token_filters.yml
@@ -127,6 +127,67 @@
     - match:  { tokens.2.token: brown }
     - match:  { tokens.3.token: fox }
 
+    - do:
+        indices.analyze:
+          body:
+            text: 'text1 #text2'
+            tokenizer: whitespace
+            filter:
+              - type: word_delimiter
+                split_on_numerics: false
+                type_table:
+                  - "\\u0023 => ALPHANUM"
+    - length: { tokens: 2 }
+    - match: { tokens.0.token: text1 }
+    - match: { tokens.0.start_offset: 0 }
+    - match: { tokens.0.end_offset: 5 }
+    - match: { tokens.0.position: 0 }
+    - match: { tokens.1.token: "#text2" }
+    - match: { tokens.1.start_offset: 6 }
+    - match: { tokens.1.end_offset: 12 }
+    - match: { tokens.1.position: 1 }
+
+    - do:
+        indices.analyze:
+          body:
+            text: 'text1 #text2'
+            tokenizer: whitespace
+            filter:
+              - type: word_delimiter
+                split_on_numerics: false
+                type_table:
+                  - "# => ALPHANUM"
+    - length: { tokens: 2 }
+    - match: { tokens.0.token: text1 }
+    - match: { tokens.0.start_offset: 0 }
+    - match: { tokens.0.end_offset: 5 }
+    - match: { tokens.0.position: 0 }
+    - match: { tokens.1.token: "#text2" }
+    - match: { tokens.1.start_offset: 6 }
+    - match: { tokens.1.end_offset: 12 }
+    - match: { tokens.1.position: 1 }
+
+    - do:
+        indices.analyze:
+          body:
+            text: 'text1 #text2'
+            tokenizer: whitespace
+            filter:
+              - type: word_delimiter
+                split_on_numerics: false
+                type_table:
+                  - "# => ALPHANUM"
+                  - "@ => ALPHANUM"
+    - length: { tokens: 2 }
+    - match: { tokens.0.token: text1 }
+    - match: { tokens.0.start_offset: 0 }
+    - match: { tokens.0.end_offset: 5 }
+    - match: { tokens.0.position: 0 }
+    - match: { tokens.1.token: "#text2" }
+    - match: { tokens.1.start_offset: 6 }
+    - match: { tokens.1.end_offset: 12 }
+    - match: { tokens.1.position: 1 }
+
 ---
 "word_delimiter_graph":
     - do:
@@ -230,6 +291,67 @@
     - match:  { detail.tokenfilters.0.tokens.5.start_offset: 16 }
     - match:  { detail.tokenfilters.0.tokens.5.end_offset: 19 }
     - match:  { detail.tokenfilters.0.tokens.5.position: 5 }
+
+    - do:
+        indices.analyze:
+          body:
+            text: 'text1 #text2'
+            tokenizer: whitespace
+            filter:
+              - type: word_delimiter_graph
+                split_on_numerics: false
+                type_table:
+                  - "\\u0023 => ALPHANUM"
+    - length: { tokens: 2 }
+    - match: { tokens.0.token: text1 }
+    - match: { tokens.0.start_offset: 0 }
+    - match: { tokens.0.end_offset: 5 }
+    - match: { tokens.0.position: 0 }
+    - match: { tokens.1.token: "#text2" }
+    - match: { tokens.1.start_offset: 6 }
+    - match: { tokens.1.end_offset: 12 }
+    - match: { tokens.1.position: 1 }
+
+    - do:
+        indices.analyze:
+          body:
+            text: 'text1 #text2'
+            tokenizer: whitespace
+            filter:
+              - type: word_delimiter_graph
+                split_on_numerics: false
+                type_table:
+                  - "# => ALPHANUM"
+    - length: { tokens: 2 }
+    - match: { tokens.0.token: text1 }
+    - match: { tokens.0.start_offset: 0 }
+    - match: { tokens.0.end_offset: 5 }
+    - match: { tokens.0.position: 0 }
+    - match: { tokens.1.token: "#text2" }
+    - match: { tokens.1.start_offset: 6 }
+    - match: { tokens.1.end_offset: 12 }
+    - match: { tokens.1.position: 1 }
+
+    - do:
+        indices.analyze:
+          body:
+            text: 'text1 #text2'
+            tokenizer: whitespace
+            filter:
+              - type: word_delimiter_graph
+                split_on_numerics: false
+                type_table:
+                  - "# => ALPHANUM"
+                  - "@ => ALPHANUM"
+    - length: { tokens: 2 }
+    - match: { tokens.0.token: text1 }
+    - match: { tokens.0.start_offset: 0 }
+    - match: { tokens.0.end_offset: 5 }
+    - match: { tokens.0.position: 0 }
+    - match: { tokens.1.token: "#text2" }
+    - match: { tokens.1.start_offset: 6 }
+    - match: { tokens.1.end_offset: 12 }
+    - match: { tokens.1.position: 1 }
 
 ---
 "unique":

--- a/modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/50_char_filters.yml
+++ b/modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/50_char_filters.yml
@@ -69,6 +69,7 @@
           char_filter:
             - type: mapping
               mappings:
+                - "# This is a comment"
                 - "# => _hashsign_"
                 - "@ => _atsign_"
   - length: { tokens: 3 }

--- a/server/src/main/java/org/opensearch/index/analysis/Analysis.java
+++ b/server/src/main/java/org/opensearch/index/analysis/Analysis.java
@@ -87,6 +87,7 @@ import java.util.List;
 import java.util.Locale;
 import java.util.Map;
 import java.util.Set;
+import java.util.regex.Pattern;
 
 import static java.util.Collections.unmodifiableMap;
 
@@ -97,6 +98,9 @@ import static java.util.Collections.unmodifiableMap;
  */
 public class Analysis {
     private static final Logger LOGGER = LogManager.getLogger(Analysis.class);
+
+    // Regular expression to support hashtag tokenization
+    private static Pattern hashTagRulePattern = Pattern.compile("^\\s*#\\s*=>");
 
     public static CharArraySet parseStemExclusion(Settings settings, CharArraySet defaultStemExclusion) {
         String value = settings.get("stem_exclusion");
@@ -222,16 +226,6 @@ public class Analysis {
         return parseWordList(env, settings, settingPrefix + "_path", settingPrefix, parser);
     }
 
-    public static <T> List<T> parseWordList(
-        Environment env,
-        Settings settings,
-        String settingPrefix,
-        CustomMappingRuleParser<T> parser,
-        boolean removeComments
-    ) {
-        return parseWordList(env, settings, settingPrefix + "_path", settingPrefix, parser, removeComments);
-    }
-
     /**
      * Parses a list of words from the specified settings or from a file, with the given parser.
      *
@@ -247,17 +241,6 @@ public class Analysis {
         String settingList,
         CustomMappingRuleParser<T> parser
     ) {
-        return parseWordList(env, settings, settingPath, settingList, parser, true);
-    }
-
-    public static <T> List<T> parseWordList(
-        Environment env,
-        Settings settings,
-        String settingPath,
-        String settingList,
-        CustomMappingRuleParser<T> parser,
-        boolean removeComments
-    ) {
         List<String> words = getWordList(env, settings, settingPath, settingList);
         if (words == null) {
             return null;
@@ -266,7 +249,7 @@ public class Analysis {
         int lineNum = 0;
         for (String word : words) {
             lineNum++;
-            if (removeComments == false || word.startsWith("#") == false) {
+            if (word.startsWith("#") == false || hashTagRulePattern.matcher(word).find()) {
                 try {
                     rules.add(parser.apply(word));
                 } catch (RuntimeException ex) {

--- a/server/src/main/java/org/opensearch/index/analysis/Analysis.java
+++ b/server/src/main/java/org/opensearch/index/analysis/Analysis.java
@@ -100,7 +100,7 @@ public class Analysis {
     private static final Logger LOGGER = LogManager.getLogger(Analysis.class);
 
     // Regular expression to support hashtag tokenization
-    private static Pattern hashTagRulePattern = Pattern.compile("^\\s*#\\s*=>");
+    private static final Pattern HASH_TAG_RULE_PATTERN = Pattern.compile("^\\s*#\\s*=>");
 
     public static CharArraySet parseStemExclusion(Settings settings, CharArraySet defaultStemExclusion) {
         String value = settings.get("stem_exclusion");
@@ -249,7 +249,7 @@ public class Analysis {
         int lineNum = 0;
         for (String word : words) {
             lineNum++;
-            if (word.startsWith("#") == false || hashTagRulePattern.matcher(word).find()) {
+            if (word.startsWith("#") == false || HASH_TAG_RULE_PATTERN.matcher(word).find() == true) {
                 try {
                     rules.add(parser.apply(word));
                 } catch (RuntimeException ex) {

